### PR TITLE
Unique message id mapper

### PR DESCRIPTION
**Pull request by aduprat**
Rework of #438, base on #442

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+
+            return cassandraAsyncExecutor.executeVoid(boundStatement);
+
+        } catch (IOException e) {
+            throw new MailboxException("Error saving mail", e);
+        }
+    }
+
+    private void setTextualLineCount(BoundStatement boundStatement, Long textualLineCount) {
+        if (Optional.ofNullable(textualLineCount).isPresent()) {
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java]:
```
textualLineCount
    .map(value -> boundStatement.setLong(TEXTUAL_LINE_COUNT, textualLineCount))
    .orElseGet(() ->boundStatement.setToNull(TEXTUAL_LINE_COUNT));
```

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+import org.junit.Rule;
+import org.junit.rules.ExpectedException;
+import org.xenei.junit.contract.Contract;
+import org.xenei.junit.contract.ContractTest;
+import org.xenei.junit.contract.IProducer;
+
+import com.google.common.collect.ImmutableList;
+
+/**
+ * Generic purpose tests for your implementation MailboxMapper.
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: It's no more mailboxMapper.

It looks like unacurate comment ;-)

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        assertThat(messages).containsOnly(message1, message4, message3);
+    }
+
+    @ContractTest
+    public void findMailboxesShouldReturnEmptyWhenMessageDoesntExist() throws MailboxException {
+        assertThat(sut.findMailboxes(mapperProvider.generateMessageId())).isEmpty();
+    }
+
+    @ContractTest
+    public void findMailboxesShouldOneMailboxWhenMessageExistsInOneMailbox() throws MailboxException {
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java]: Should'Return'One

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/FlagsExtractor.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    public Flags getFlags() {
+        Flags flags = new Flags();
+        for (String flag : CassandraMessageTable.Flag.ALL) {
+            if (row.getBool(flag)) {
+                flags.add(CassandraMessageTable.Flag.JAVAX_MAIL_FLAG.get(flag));
+            }
+        }
+        row.getSet(CassandraMessageTable.Flag.USER_FLAGS, String.class)
+            .stream()
+            .forEach(flags::add);
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/FlagsExtractor.java]: With a reduce to avoid side effects ?

**Comment by aduprat** [comment]
@chibenwa please be patient, i've not yet published this review; it's a building test.
But the final review will look similar.

**Comment by aduprat** [comment]
test this please

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/CassandraMessageId.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    public static class Factory implements MessageId.Factory {
+
+        @Override
+        public CassandraMessageId generate() {
+            return of(UUIDs.timeBased());
+        }
+    }
+
+    public static CassandraMessageId of(UUID uuid) {
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/CassandraMessageId.java]: I would put that in the factory interface

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraMessageIdTest.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.mailbox.cassandra;
+
+import org.junit.Test;
+
+import nl.jqno.equalsverifier.EqualsVerifier;
+
+public class CassandraMessageIdTest {
```
Comment (mbaechler) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraMessageIdTest.java]: you need more tests on CassandraMessageId

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdToImapUidDAOTest.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+        testee.delete(messageId, mailboxId).join();
+
+        ComposedMessageId expectedComposedMessageId = new ComposedMessageId(mailboxId2, messageId, messageUid2);
+        Stream<ComposedMessageId> messages = testee.retrieve(messageId, Optional.empty()).join();
+        assertThat(messages.collect(Guavate.toImmutableList())).containsOnly(expectedComposedMessageId);
+    }
+
+    @Test
+    public void insertShouldInsert() {
```
Comment (mbaechler) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdToImapUidDAOTest.java]: seriously ?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdToImapUidDAOTest.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+        testee.insert(messageId, mailboxId, messageUid).join();
+
+        ComposedMessageId expectedComposedMessageId = new ComposedMessageId(mailboxId, messageId, messageUid);
+        Stream<ComposedMessageId> messages = testee.retrieve(messageId, Optional.of(mailboxId)).join();
+        assertThat(messages.collect(Guavate.toImmutableList())).containsOnly(expectedComposedMessageId);
+    }
+
+    @Test
+    public void retrieveShouldRetrieveWhenKeyMatches() {
```
Comment (mbaechler) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdToImapUidDAOTest.java]: hmm

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdDAOTest.java [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+        CassandraMessageId messageId = messageIdFactory.generate();
+        CassandraMessageId messageId2 = messageIdFactory.generate();
+        CassandraMessageId messageId3 = messageIdFactory.generate();
+        CassandraMessageId messageId4 = messageIdFactory.generate();
+        CassandraId mailboxId = CassandraId.timeBased();
+        MessageUid messageUid = MessageUid.of(1);
+        MessageUid messageUid2 = MessageUid.of(2);
+        MessageUid messageUid3 = MessageUid.of(3);
+        MessageUid messageUid4 = MessageUid.of(4);
+        testee.insert(mailboxId, messageUid, messageId).join();
```
Comment (mbaechler) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdDAOTest.java]: the point is to join on the Future<List>> not to join on individual elements

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+                .setLong(MOD_SEQ, message.getModSeq())
+                .setUUID(MESSAGE_ID, messageId.get())
+                .setLong(MOD_SEQ, oldModSeq))
+            .thenApply(optional -> optional
+                    .map(row -> row.getBool(CassandraConstants.LIGHTWEIGHT_TRANSACTION_APPLIED))
+                    .orElse(false));
+    }
+    
+    public List<MailboxMessage> retrieveMessages(List<CassandraMessageId> messageIds, FetchType fetchType, Optional<Integer> limit, Function<List<AttachmentId>, List<Attachment>> attachmentsFunction) {
+        return CassandraUtils.convertToStream(cassandraAsyncExecutor.execute(buildSelectQueryWithLimit(buildQuery(messageIds, fetchType), limit)).join())
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java]: please use some line breaks to make it more readable

Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java]: why do you need to join ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+    private final CassandraMessageIdToImapUidDAO messageIdToImapUidDAO;
+    private final PreparedStatement insert;
+    private final PreparedStatement update;
+    private final PreparedStatement delete;
+
+    @Inject
+    public CassandraMessageDAO(Session session, CassandraTypesProvider typesProvider, CassandraMessageIdToImapUidDAO messageIdToImapUidDAO) {
+        this.cassandraAsyncExecutor = new CassandraAsyncExecutor(session);
+        this.typesProvider = typesProvider;
+        this.messageIdToImapUidDAO = messageIdToImapUidDAO;
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java]: why do you need that, it should be used in CassandraMessageMapper, don't you think ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 12]
```diff
@@ -174,53 +120,53 @@ public long countUnseenMessagesInMailbox(Mailbox mailbox) throws MailboxExceptio
...
+
         decrementCount(mailboxId);
         if (!message.isSeen()) {
             decrementUnseen(mailboxId);
         }
     }
 
+    private Optional<CassandraMessageId> retrieveMessageId(CassandraId mailboxId, MailboxMessage message) {
+        return messageIdDAO.retrieve(mailboxId, message.getUid())
+                .join();
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: no join !

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 13]
```diff
@@ -174,53 +120,53 @@ public long countUnseenMessagesInMailbox(Mailbox mailbox) throws MailboxExceptio
...
     @Override
     public List<MessageUid> findRecentMessageUidsInMailbox(Mailbox mailbox) throws MailboxException {
         CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
-        return CassandraUtils.convertToStream(session.execute(selectAll(mailboxId, FetchType.Metadata).and((eq(RECENT, true)))))
-            .map((row) -> row.getLong(IMAP_UID))
-            .map(MessageUid::of)
+        return messageDAO.retrieveMessages(messageIdDAO.retrieveMessageIds(mailboxId, MessageRange.all()), FetchType.Metadata, Optional.empty(), attachmentMapper::getAttachments).stream()
+            .filter(MailboxMessage::isRecent)
+            .flatMap(message -> imapUidDAO.retrieve((CassandraMessageId) message.getMessageId(), Optional.ofNullable(mailboxId)).join())
+            .map(ComposedMessageId::getUid)
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: no join here

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraMailboxSessionMapperFactoryTest.java [hunk 14]
```diff
@@ -63,8 +62,7 @@ public void beforeMethod() {
     public void testCreateMessageMapper() throws Exception {
         LOG.info("createMessageMapper");
         MailboxSession session = null;
-        DefaultMessageId.Factory messageIdFactory = new DefaultMessageId.Factory();
-        CassandraMailboxSessionMapperFactory instance = new CassandraMailboxSessionMapperFactory(null, null, null, null, messageIdFactory);
+        CassandraMailboxSessionMapperFactory instance = new CassandraMailboxSessionMapperFactory(null, null, null, null, null, null, null);
```
Comment (mbaechler) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraMailboxSessionMapperFactoryTest.java]: please don't stack nulls

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java (hunk 11 above)
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java]: Removed in next commits.

**Comment by aduprat** [comment]
Refactor in #463

**Pull request closed without merging**